Add Data Mutation Guide (#994)
solidjs/solid-docs: src/routes/solid-start/guides/actions-and-mutations.mdx

1. Create an action with a unique name
2. Pass your action to the `<form>` element in the `action` prop
3. Make sure the `<form>` element uses the `post` method

```tsx {3-9} {13} title="src/routes/index.tsx"
import { action } from "@solidjs/router";

const addPost = action(async (formData: FormData) => {
	const title = formData.get("title");
	await fetch("https://my-api.com/posts", {
		method: "POST",
		body: JSON.stringify({ title }),
	});
}, "addPost");

export default function Page() {
	return (
		<form action={addPost} method="post">
			<input name="title" />
			<button>Add Post</button>
		</form>
	);
}
```

When invoked in a form, the action automatically receives the [`FormData`](https://developer.mozilla.org/en-US/docs/Web/API/FormData/FormData) object. You can extract the field data using the native `FormData` methods.

If your action accepts typed data, you can use the `with` method to pass custom data to the action:

```tsx {14} title="src/routes/index.tsx"
import { createSignal } from "solid-js";
import { action } from "@solidjs/router";

> const addPost = action(async (title: string) => {
> 	await fetch("https://my-api.com/posts", {
> 		method: "POST",
> 		body: JSON.stringify({ title }),
> 	});
> }, "addPost");
> 
> export default function Page() {
> 	const [title, setTitle] = createSignal("");
> 	return (
> 		<form action={addPost.with(title())} method="post">
> 			<input value={title()} onChange={(e) => setTitle(e.target.value)} />
> 			<button>Add Post</button>
> 		</form>
> 	);
> }
> ```
> 
> Note that the action takes a `string` as its parameter rather than a `FormData`.

Reviewer: ```suggestion
const addPost = action(async (title: string, formData: FormData) => {
	await fetch("https://my-api.com/posts", {
		method: "POST",
		body: JSON.stringify({ title }),
	});
}, "addPost");

export default function Page() {
	const [title, setTitle] = createSignal("");
	return (
		<form action={addPost.with(title())} method="post">
			<input value={title()} onChange={(e) => setTitle(e.target.value)} />
			<button>Add Post</button>
		</form>
	);
}
```

Note that the action now takes a `string` as its first parameter and the `FormData` as its second.
```

Author: I tried not to go into details of APIs in this guide. I believe that this content is better suited for the API Reference.

Reviewer: I understand, but saying
> the action takes a `string` as its parameter rather than a `FormData`

just felt a little misleading, it gives off the impression that the formdata cannot be used anymore if you switch to the `.with` form.

Author: How about this:

> Note that the action takes a `string` as its **first** parameter rather than a `FormData`.

This doesn't imply that `FormData` is not accessible inside the action, and doesn't go too much API details.

I don't want to mention the `FormData` is passed as the second argument because it's not related to the example.

Author: Do you have a practical example of using `FormData` inside an action when using `with()`? If there is a common pattern that I'm not aware of, I'm open to update this particular example.

Reviewer: I think this wording makes more sense to me now, I'm good with it.

Also [here's an example](https://www.jacobparis.com/content/remix-multiple-actions#multiple-actions-from-a-single-form) that shows one use case with Remix where we would need both arguments

Instead of the hidden input like in this example, we would pass the tweet id using `.with`, and use the formdata to determine which action was taken.

Author: I completely rewrote this section. Instead of focusing solely on typed data, I shifted the emphasis to passing additional data. This change better represents the API, and it implicitly suggests that the additional arguments are typed, so there's no need to state that explicitly. 

What do you think about this approach?

> ## Showing loading UI
> 
> To display a loading UI while the action is being executed, you can use the [`useSubmission`](/solid-router/reference/data-apis/use-submission) primitive:
> 
> 1. Import `useSubmission` from `@solidjs/router`
> 2. Call `useSubmission` with your action
> 3. Use the `pending` property to show a loading state

Reviewer: ```suggestion
## Showing pending UI

To display a pending UI while the action is being executed, you can use the [`useSubmission`](/solid-router/reference/data-apis/use-submission) primitive:

1. Import `useSubmission` from `@solidjs/router`
2. Call `useSubmission` with your action
3. Use the `pending` property to show a pending state
```
To be consistent with the submission api, and to draw a clear distinction from loading states during data fetching. Calling this `loading` will inevitable lead someone to wonder if we can use Suspense here.

import { For, Show } from "solid-js";
import { action, useSubmission, query, createAsync } from "@solidjs/router";

const getPosts = query(async () => {
	const posts = await fetch("https://my-api.com/blog");
	return await posts.json();
}, "posts");

const addPost = action(async (formData: FormData) => {
	const title = formData.get("title");
	await fetch("https://my-api.com/posts", {
		method: "POST",
		body: JSON.stringify({ title }),
	});
}, "addPost");

export default function Page() {
	const posts = createAsync(() => getPosts());
	const submission = useSubmission(addPost);
	return (
		<main>
			<form action={addPost} method="post">
				<input name="title" />
				<button>Add Post</button>
			</form>
			<ul>
				<For each={posts()}>{(post) => <li>{post.title}</li>}</For>
				<Show when={submission.pending}>
					{submission.input?.[0]?.get("title")?.toString()}
				</Show>
			</ul>
		</main>
	);
}
```

<Callout type="info" title="Multiple Submissions">
	If you want to display optimistic UI for multiple concurrent submissions, you can use the [`useSubmissions`](/solid-router/reference/data-apis/use-submissions) primitive.
</Callout>

> ## Calling an action manually
> 
> If you don't want to use a `<form>` element, you can use the `useAction` primitive to manually call an action:

Reviewer: ```suggestion
## Calling an action programmatically 

If you don't want to use a `<form>` element, you can use the `useAction` primitive to programmatically call an action:
```

1. Import `useAction` from `@solidjs/router`
2. Call `useAction` with your action
3. Use the returned function to trigger the action

```tsx {13} {17} title="src/routes/index.tsx"
import { createSignal } from "solid-js";
import { action, useAction } from "@solidjs/router";

const addPost = action(async (title: string) => {
	await fetch("https://my-api.com/posts", {
		method: "POST",
		body: JSON.stringify({ title }),
	});
}, "addPost");

export default function Page() {
	const [title, setTitle] = createSignal("");
	const addPostAction = useAction(addPost);
	return (
		<div>
			<input value={title()} onInput={(e) => setTitle(e.target.value)} />

> 			<button onClick={() => addPostAction(title())}>Add Post</button>

Reviewer: I would recommend adding a callout at the bottom of this section that explains why form actions are preferred over `useAction` in terms of accessibility and progressive enhancement.

Author: This is better suited for a tutorial.

Like the Data Fetching guide, this guide is geared toward people who want to know *how* to do things, and not why and when.

---
title: "Actions and Mutations"
---

This guide provides practical examples of using actions to mutate data in SolidStart.

## Handling form submission

To handle [`<form>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form) submissions with an action:

> 1. Ensure the action has a unique name. [Learn more here](/solid-router/reference/data-apis/action#notes-of-form-implementation-and-ssr).

Reviewer 2: ```suggestion
1. Ensure the action has a unique name. [More information on the Action API reference](/solid-router/reference/data-apis/action#notes-of-form-implementation-and-ssr).
```

Author: I prefer mine because it's shorter, more action-oriented, and flows better. It also aligns with the style of other links in the docs (at least from what I've seen). Do you have any concerns about it?

Reviewer 3: When you link out to other pages, you need for them to be descriptive for the [sake of accessibility](https://webaim.org/techniques/hypertext/).

Author: I think neither option is fully accessible because they don’t make sense when read out of context. Although "More information on the Action API reference" is more descriptive about the destination, it’s unclear what "more" refers to.

I propose this instead:

> See the [Action API reference](https://github.com/solid-router/reference/data-apis/action#notes-of-form-implementation-and-ssr) for more information.

This combines the best aspects of both options: it’s more actionable and inviting while remaining accessible. The link also makes sense when read independently of its context.

Reviewer 3: I think that's being caught up in the details. Both options are reasonable and would work. The suggestion was to say the first option would not.

2. Pass the action to the `<form>` element using the `action` prop.
3. Ensure the `<form>` element uses the `post` method for submission.
4. Use the [`FormData`](https://developer.mozilla.org/en-US/docs/Web/API/FormData/FormData) object in the action to extract field data using the navite `FormData` methods.

<TabsCodeBlocks>
<div id="TypeScript">

```tsx {3-9} {13} title="src/routes/index.tsx"
import { action } from "@solidjs/router";

const addPost = action(async (formData: FormData) => {

> 	const title = formData.get("title");

Reviewer 2: this will raise ~~alarms~~ squiggles in the TypeScript LSP because a form value can be `string`, `number`, `File`, or `null`.

The user will either need to add a type guard or a type assertion here.
Though I generally consider type assertion a code smell, it's quite a harmless one in this specific scenario and we need to consider the focus of this code snippet.

```suggestion
	const title = formData.get("title") as string;
```

	await fetch("https://my-api.com/posts", {
		method: "POST",
		body: JSON.stringify({ title }),
	});
}, "addPost");

export default function Page() {
	return (
		<form action={addPost} method="post">
			<input name="title" />
			<button>Add Post</button>
		</form>
	);
}
```

</div>
<div id="JavaScript">

```jsx {3-9} {13} title="src/routes/index.jsx"
import { action } from "@solidjs/router";

const addPost = action(async (formData) => {
	const title = formData.get("title");
	await fetch("https://my-api.com/posts", {
		method: "POST",
		body: JSON.stringify({ title }),
	});
}, "addPost");

export default function Page() {
	return (
		<form action={addPost} method="post">
			<input name="title" />
			<button>Add Post</button>
		</form>
	);
}
```

</div>
</TabsCodeBlocks>

## Passing additional arguments

To pass additional arguments to your action, use the `with` method:

<TabsCodeBlocks>
<div id="TypeScript">

```tsx {14} title="src/routes/index.tsx"
import { action } from "@solidjs/router";

const addPost = action(async (userId: number, formData: FormData) => {

> 	const title = formData.get("title");

Reviewer 2: same as above.

```suggestion
	const title = formData.get("title") as string;
```

	const userId = 1;
	return (
		<form action={addPost.with(userId)} method="post">
			<input name="title" />
			<button>Add Post</button>
		</form>
	);
}
```

</div>
<div id="JavaScript">

```jsx {14} title="src/routes/index.jsx"
import { action } from "@solidjs/router";

const addPost = action(async (userId, formData) => {
	const title = formData.get("title");
	await fetch("https://my-api.com/posts", {
		method: "POST",
		body: JSON.stringify({ userId, title }),
	});
}, "addPost");

export default function Page() {
	const userId = 1;
	return (
		<form action={addPost.with(userId)} method="post">
			<input name="title" />
			<button>Add Post</button>
		</form>
	);
}
```

</div>
</TabsCodeBlocks>

## Showing pending UI

To display a pending UI during action execution:

1. Import [`useSubmission`](/solid-router/reference/data-apis/use-submission) from `@solidjs/router`.
2. Call `useSubmission` with your action, and use the returned `pending` property to display pending UI.

<TabsCodeBlocks>
<div id="TypeScript">

```tsx {12} {16-18} title="src/routes/index.tsx"
import { action, useSubmission } from "@solidjs/router";

const addPost = action(async (formData: FormData) => {

> 	const title = formData.get("title");

Reviewer 2: same as above.

```suggestion
	const title = formData.get("title") as string;
```

			</button>
		</form>
	);
}
```

</div>
<div id="JavaScript">

```jsx {12} {16-18} title="src/routes/index.jsx"
import { action, useSubmission } from "@solidjs/router";

const addPost = action(async (formData) => {
	const title = formData.get("title");
	await fetch("https://my-api.com/posts", {
		method: "POST",
		body: JSON.stringify({ title }),
	});
}, "addPost");

export default function Page() {
	const submission = useSubmission(addPost);
	return (
		<form action={addPost} method="post">
			<input name="title" />
			<button disabled={submission.pending}>
				{submission.pending ? "Adding..." : "Add Post"}
			</button>
		</form>
	);
}
```

</div>
</TabsCodeBlocks>

## Handling errors

To handle errors that occur within an action:

1. Import [`useSubmission`](/solid-router/reference/data-apis/use-submission) from `@solidjs/router`.
2. Call `useSubmission` with your action, and use the returned `error` property to handle the error.

<TabsCodeBlocks>
<div id="TypeScript">

```tsx {13} {16-18} title="src/routes/index.tsx"
import { Show } from "solid-js";
import { action, useSubmission } from "@solidjs/router";

const addPost = action(async (formData: FormData) => {

> 	const title = formData.get("title");

Reviewer 2: same as above.

```suggestion
	const title = formData.get("title") as string;
```

	});
}, "addPost");

export default function Page() {
	const posts = createAsync(() => getPosts());
	const submission = useSubmission(addPost);
	return (
		<main>
			<form action={addPost} method="post">
				<input name="title" />
				<button>Add Post</button>
			</form>
			<ul>
				<For each={posts()}>{(post) => <li>{post.title}</li>}</For>
				<Show when={submission.pending}>
					{submission.input?.[0]?.get("title")?.toString()}
				</Show>
			</ul>
		</main>
	);
}
```

</div>
</TabsCodeBlocks>

<Callout type="info" title="Multiple Submissions">

If you want to display optimistic UI for multiple concurrent submissions, you can use the [`useSubmissions`](/solid-router/reference/data-apis/use-submissions) primitive.

</Callout>

## Redirecting

To redirect users to a different route within an action:

1. Import [`redirect`](/solid-router/reference/response-helpers/redirect) from `@solidjs/router`.
2. Call `redirect` with the route you want to navigate to, and throw its response.

<TabsCodeBlocks>
<div id="TypeScript">

```tsx {10} title="src/routes/index.tsx"
import { action, redirect } from "@solidjs/router";

const addPost = action(async (formData: FormData) => {

> 	const title = formData.get("title");

Reviewer 2: same as above.

```suggestion
	const title = formData.get("title") as string;
```

	return (
		<form action={addPost} method="post">
			<input name="title" />
			<button>Add Post</button>
		</form>
	);
}
```

</div>
<div id="JavaScript">

```jsx {10} title="src/routes/index.jsx"
import { action, redirect } from "@solidjs/router";

const addPost = action(async (formData) => {
	const title = formData.get("title");
	const response = await fetch("https://my-api.com/posts", {
		method: "POST",
		body: JSON.stringify({ title }),
	});
	const post = await response.json();
	throw redirect(`/posts/${post.id}`);
}, "addPost");

export default function Page() {
	return (
		<form action={addPost} method="post">
			<input name="title" />
			<button>Add Post</button>
		</form>
	);
}
```

</div>
</TabsCodeBlocks>

## Using a database or an ORM

To safely interact with your database or ORM in an action, ensure it's server-only by adding [`"use server"`](/solid-start/reference/server/use-server) as the first line of your action:

<TabsCodeBlocks>
<div id="TypeScript">

```tsx {5} title="src/routes/index.tsx"
import { action } from "@solidjs/router";
import { db } from "~/lib/db";

const addPost = action(async (formData: FormData) => {
	"use server";

> 	const title = formData.get("title");

Reviewer 2: same as above.

```suggestion
	const title = formData.get("title") as string;
```
